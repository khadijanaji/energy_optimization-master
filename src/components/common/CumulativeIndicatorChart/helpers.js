/* eslint-disable no-magic-numbers, react/no-multi-comp, max-len, max-lines */
// @flow


import moment from "moment";
import __, {isNumber, range, isNull} from "lodash";
import * as React from "react";

import strings from "./strings";
import loadingIcon from "../../../img/loading.svg";

export const recordBeginning = 7;
export const dayHours = 24;
export const precision = 3;

export const TIMELINETYPE = {
  "DAILY": "DAILY",
  "MONTHLY": "MONTHLY",
  "YEARLY": "YEARLY",
  "CUSTOM_MONTHLY": "CUSTOM_MONTHLY",
  "CUSTOM_YEARLY": "CUSTOM_YEARLY"
};

export const getYesterday = (): moment => moment().subtract(1, "days");

export const getMonthsOfThisYear = (date: moment): Array<string> => {
  const thisMonth = date.month() + 1;
  const tmp = date.clone();
  return range(0, thisMonth).map((ind) => (tmp.month(ind).format("MMM")));
};

export const getMonthsOfRange = (fromDate: moment, toDate: moment): Array<string> => {
  const fromMonth = fromDate.month() + 1;
  const toMonth = toDate.month() + 1;
  const tmp = fromDate.clone();
  return range(fromMonth, toMonth).map((index) => (tmp.month(index).format("MMM")));
};

export const getMonthsOfRanges = (fromDate: moment, toDate: moment): Array<string> => {
  const output = [];
  const dateCursor = fromDate.clone();
  while (dateCursor.isSameOrBefore(toDate)) {
    output.push(dateCursor.format("MMM"));
    dateCursor.add(1, "days");
  }
  return output;
};

export const getFirstDayOfThisMonth = (yesterday: moment): moment => (yesterday.
  clone().
  startOf("month"));

export const getDaysOfMonth = (dayOfMonth: moment): Array<string> => {
  const firstDay = getFirstDayOfThisMonth(dayOfMonth);
  const dayNumber = dayOfMonth.date();
  return range(0, dayNumber).map((ind) => (firstDay.
    clone().
    add(ind, "days").
    format("DD")));
};

export const getDaysOfMonthBetween = (fromDate: moment, toDate: moment): Array<string> => {
  const output = [];
  const dateCursor = fromDate.clone();
  while (dateCursor.isSameOrBefore(toDate)) {
    output.push(dateCursor.format("DD"));
    dateCursor.add(1, "days");
  }
  return output;
};

export const getDatesBetween = (fromDate: moment, toDate: moment): Array<string> => {
  const output = [];
  const dateCursor = fromDate.clone();
  while (dateCursor.isSameOrBefore(toDate)) {
    output.push(dateCursor.format("DD/MM"));
    dateCursor.add(1, "days");
  }
  return output;
};

export const getTargetsFromRecordBeginningToCurrentMonth = (yearlyTarget: number, currentMonth: number) => {
  const monthsInYear = 12;
  const monthlyTarget = yearlyTarget / monthsInYear;

  return __.range(1, currentMonth + 2).map((elem) => (Math.round(elem * monthlyTarget)));
};

export const getTargetsFromRecordBeginningToCurrentDay = (monthlyTarget: number, currentDayOfMonth: moment) => {
  const currentDay = currentDayOfMonth.date();
  const daysInmonth = currentDayOfMonth.daysInMonth();
  const dailyTarget = monthlyTarget / daysInmonth;

  return __.range(1, currentDay + 1).map((elem) => (Math.round(elem * dailyTarget)));
};

export const getTodayProductionHours = () => {
  const now = moment().hour();
  return (now >= recordBeginning)
    ? (now - recordBeginning + 1)
    : (dayHours - recordBeginning + now);
};

export const getHoursFromRecordBeginningToCurrentHour = () => {
  const now = moment().hour();
  if (now >= recordBeginning) {
    return __.
      range(recordBeginning, now + 1).
      map((elem) => `${elem} h`);
  }
  return __.
    range(recordBeginning, dayHours).
    concat(__.range(0, now + 1)).
    map((elem) => `${elem} h`);
};

export const getTargetsFromRecordBeginningToCurrentHour = (target: number) => {
  const productionHours = getTodayProductionHours();
  const hourlyTarget = target / dayHours;
  const targets = __.range(0, productionHours).map((elem) => (Math.round(elem * hourlyTarget)));
  targets[0] = 0;
  return targets;
};

export const getActualTargetMonth = (monthlyTarget: number) => {
  const currentDay = getYesterday().date();
  const daysInmonth = getYesterday().daysInMonth();
  if (monthlyTarget) {
    return (monthlyTarget / daysInmonth) * currentDay;
  }
  return (<span className="cumulative-indicator-chart__container__value__UNAVAILABLE">{"Donnée non disponible"}</span>);
};

export const getActualTarget = (target: number) => {
  const productionHours = getTodayProductionHours();
  if (target) {
    const ratio = (productionHours - 1) / dayHours;
    return target * ratio;
  }
  return (<span className="cumulative-indicator-chart__container__value__UNAVAILABLE">{"Donnée non disponible"}</span>);
};

export const getGapBetweenActualAndTargetValue = (target: number, value: number) => {
  if (target) {
    return value - target;
  }
  return (<span className="cumulative-indicator-chart__container__value__UNAVAILABLE">{"Donnée non disponible"}</span>);
};

export const getGapPercent =
    (target: number, value: number) => {
      if (target) {
        return (100.0 * parseFloat(getGapBetweenActualAndTargetValue(target, value)) / target);
      }
      return (<span className="cumulative-indicator-chart__container__value__UNAVAILABLE">
        {"Donnée non disponible"}
      </span>);
    };

type mapChartDataType = {
  title: string,
  subtitle: string,
  unit: string,
  target?: Array<number>,
  values: Array<number>,
  timeLineType: string,
  fromDate?: moment,
  toDate?: moment
};

export const formatValues = (values: Array<number>) => {
  if (values) {
    return values.map((value) => value && parseFloat(Math.round(value * 1000) / 1000));
  }
  return values;
};

export const formatYearlyValues = (values: Array<number>, names: Array<string>) => {
  if (values) {
    return values.map((value, index) => ({
      "name": names[index],
      "y": value && parseFloat(Math.round(value * 1000) / 1000)
    }));
  }
  return values;
};

export const mapValuesToCumulativeValues =
    (values: Array<number>): Array<number> => {
      let sum = 0;
      return values && values.map((value) => (sum += value));
    };

export const mapDailyChartData = ({title, subtitle, unit, target, values}: mapChartDataType) => {
  const categories = getHoursFromRecordBeginningToCurrentHour();
  const data = target && getTargetsFromRecordBeginningToCurrentHour(target[0]);
  const allValues = values && [].concat(0, ...values);
  return {
    "chart": {
      "type": "spline",
      "zoomType": "xy",
      "pinchType": "xy"
    },
    "title": {"text": title},
    "subtitle": {"text": subtitle},
    "xAxis": {categories},
    "yAxis": {"title": {"text": unit}},
    "plotOptions": {
      "line": {
        "dataLabels": {"enabled": true},
        "enableMouseTracking": false
      }
    },
    "tooltip": {
      "formatter" () {
        return `<span>${this.x.toString()}</span> : <b style="color:${this.color}">${this.y.toString()}</b>`;
      },
      "useHTML": true
    },
    "series": [
      {
        "name": strings.target,
        data,
        "dashStyle": "Dot",
        "color": "#A8FBA8"
      },
      {
        "name": strings.actuals,
        "data": formatValues(allValues)
      }
    ]
  };
};

export const mapMonthlyChartData = ({title, subtitle, unit, target, values}: mapChartDataType) => {
  const categories = getDaysOfMonth(getYesterday());
  const data = target && getTargetsFromRecordBeginningToCurrentDay(target[0], getYesterday());
  return {
    "chart": {
      "type": "spline",
      "zoomType": "xy",
      "pinchType": "xy"
    },
    "title": {"text": title},
    "subtitle": {"text": subtitle},
    "xAxis": {categories},
    "yAxis": {"title": {"text": unit}},
    "plotOptions": {
      "line": {
        "dataLabels": {"enabled": true},
        "enableMouseTracking": false
      }
    },
    "tooltip": {
      "formatter" () {
        return `<span>${this.x.toString()}</span> : <b style="color:${this.color}">${this.y.toString()}</b>`;
      },
      "useHTML": true
    },
    "series": [
      {
        "name": strings.target,
        data,
        "dashStyle": "Dot",
        "color": "#A8FBA8"
      },
      {
        "name": strings.actuals,
        "data": formatValues(mapValuesToCumulativeValues(values))
      }
    ]
  };
};

const replaceValuesWithNull = (values: Array<number>) => values.map((value) => {
  if (isNull(value)) {
    return 0;
  }
  return null;
});

const formatTooltipValue = (color: string, value: string) => {
  if (color === "red") {
    return "Valeur non disponible";
  }
  return value;
};

const rangeHistoricTooltip = (dates: Array<string>) => ({
  "formatter" () {
    return `<span>${dates[this.series.data.indexOf(this.point)]}</span> : <b style="color:${this.color}">${formatTooltipValue(this.color, this.y.toString())}</b>`;
  },
  "useHTML": true
});

export const mapMonthlyRangeChartData = ({title, subtitle, unit, values, fromDate, toDate, target}: mapChartDataType) => {
  const categories = getDaysOfMonthBetween(fromDate, toDate);
  const dates = getDatesBetween(fromDate, toDate);
  return {
    "chart": {
      "type": "spline",
      "zoomType": "xy",
      "pinchType": "xy"
    },
    "title": {"text": title},
    "subtitle": {"text": subtitle},
    "xAxis": {
      categories,
      "gridLineWidth": "1",
      "lineWidth": 1,
      "tickmarkPlacement": "on"
    },
    "yAxis": {"title": {"text": unit}},
    "plotOptions": {
      "line": {
        "dataLabels": {"enabled": true},
        "enableMouseTracking": false
      }
    },
    "tooltip": rangeHistoricTooltip(dates),
    "series": [
      {
        "name": "Valeurs non disponible",
        "data": replaceValuesWithNull(values),
        "color": "red"
      },
      {
        "name": strings.actuals,
        "data": formatValues(values)
      },
      {
        "name": strings.target,
        "data": target,
        "dashStyle": "ShortDash",
        "color": "#A8FBA8"
      }
    ]
  };
};

const positionOfUniqueMonths = (months: string[]) => {
  const result = [0];
  for (let index = 1; index < months.length; index += 1) {
    if (months[index] !== months[index - 1]) {
      result.push(index);
    }
  }
  return result;
};

export const mapYearlyRangeChartData = ({title, subtitle, unit, values, fromDate, toDate, target}: mapChartDataType) => {
  const categories = getMonthsOfRanges(fromDate, toDate);
  const dates = getDatesBetween(fromDate, toDate);
  return {
    "chart": {
      "type": "spline",
      "zoomType": "xy",
      "pinchType": "xy"
    },
    "title": {"text": title},
    "subtitle": {"text": subtitle},
    "xAxis": {
      categories,
      "tickPositions": positionOfUniqueMonths(categories),
      "gridLineWidth": "1",
      "lineWidth": 1,
      "tickmarkPlacement": "on",
      "isCustomYearly": true
    },
    "yAxis": {"title": {"text": unit}},
    "plotOptions": {
      "line": {
        "dataLabels": {"enabled": true},
        "enableMouseTracking": false
      }
    },
    "tooltip": rangeHistoricTooltip(dates),
    "series": [
      {
        "name": "Valeurs non disponible",
        "data": replaceValuesWithNull(values),
        "color": "red",
        "lineWidth": 0
      },
      {
        "name": strings.actuals,
        "data": formatYearlyValues(values, categories)
      },
      {
        "name": strings.target,
        "data": target,
        "dashStyle": "ShortDash",
        "color": "#A8FBA8"
      }
    ]
  };
};

export const mapYearlyChartData = ({title, subtitle, unit, target, values}: mapChartDataType) => {
  const categories = getMonthsOfThisYear(getYesterday());
  return {
    "chart": {
      "type": "column",
      "zoomType": "xy",
      "pinchType": "xy"
    },
    "title": {"text": title},
    "subtitle": {"text": subtitle},
    "xAxis": {categories},
    "yAxis": {"title": {"text": unit}},
    "plotOptions": {
      "line": {
        "dataLabels": {"enabled": true},
        "enableMouseTracking": false
      }
    },
    "tooltip": {
      "formatter" () {
        return `<span>${this.x.toString()}</span> : <b style="color:${this.color}">${this.y.toString()}</b>`;
      },
      "useHTML": true
    },
    "series": [
      {
        "name": strings.target,
        "data": target,
        "dashStyle": "Dot",
        "color": "#A8FBA8"
      },
      {
        "name": strings.actuals,
        "data": formatValues(values)
      }
    ]
  };
};

export const mapChartData = (mapchartDataParams: mapChartDataType) => {
  switch (mapchartDataParams.timeLineType) {
  case TIMELINETYPE.MONTHLY:
    return mapMonthlyChartData(mapchartDataParams);
  case TIMELINETYPE.YEARLY:
    return mapYearlyChartData(mapchartDataParams);
  case TIMELINETYPE.CUSTOM_MONTHLY:
    return mapMonthlyRangeChartData(mapchartDataParams);
  case TIMELINETYPE.CUSTOM_YEARLY:
    return mapYearlyRangeChartData(mapchartDataParams);
  case TIMELINETYPE.DAILY:
  default:
    return mapDailyChartData(mapchartDataParams);
  }
};

export const renderValue = (value: any, status: string) => {
  if (status === "Loading") {
    return (
      <img
        alt={"chargement"}
        className="category-indicator-item__loading"
        src={loadingIcon}
      />
    );
  }
  if (isNumber(value)) {
    return (parseFloat(value).toFixed(precision));
  }
  return (<span className="cumulative-indicator-chart__container__value__UNAVAILABLE">{"Donnée non disponible"}</span>);
};

export const getGravityClassName = (actualTarget: number, value: number) => (value && ((value >= actualTarget)
  ? "cumulative-indicator-chart__container__value__OK"
  : "cumulative-indicator-chart__container__value__DANGER"));
