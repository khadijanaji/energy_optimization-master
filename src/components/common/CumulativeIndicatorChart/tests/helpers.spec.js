/* eslint-disable max-lines */
/* eslint-disable no-magic-numbers,global-require */
/* globals describe it expect beforeAll beforeEach */

import {TIMELINETYPE} from "../helpers";

let helpers = null;
let moment = null;

const value = 123.45;
const target = [150.20];
const actualTarget1 = [140];
const actualTarget2 = [81];
const values = [
  10,
  30,
  50,
  78.5,
  96.17,
  123.45
];
const monthlyTarget = [129000];
const yearlyTarget = [
  1350,
  3000,
  5012,
  7850,
  9617,
  12345
];

describe("Component", () => {
  describe("CumulativeIndicatorChart Helpers", () => {
    beforeAll(() => {
      jest.resetModules();
      jest.doMock("moment", () => () => ({"hour": () => 19}));
      helpers = require("../helpers");
    });
    describe("renderValue Render", () => {
      it("With correct data, should render value", () => {
        const wrapper = helpers.renderValue(value, "Success");
        expect(wrapper).toMatchSnapshot();
      });

      it("With loading status, should render loader", () => {
        const wrapper = helpers.renderValue(value, "Loading");
        expect(wrapper).toMatchSnapshot();
      });

      it("With incorrect value, should render N/A", () => {
        const wrapper = helpers.renderValue("123.45", "Success");
        expect(wrapper).toMatchSnapshot();
      });
    });
    describe("mapDailyChartData Mapper", () => {
      it("With Correct params, should match snapshot", () => {
        const wrapper = helpers.mapDailyChartData({
          "title": "title",
          "subtitle": "subtitle",
          "unit": "unit",
          target,
          values
        });
        expect(wrapper).toMatchSnapshot();
      });
    });
    describe("getTodayProductionHours Helper", () => {
      it("With Correct params, should match snapshot", () => {
        const wrapper = helpers.getTodayProductionHours();
        expect(wrapper).toMatchSnapshot();
      });
    });
    describe("getTargetsFromRecordBeginningToCurrentHour Helper", () => {
      it("With Correct params, should match snapshot", () => {
        const wrapper = helpers.getTargetsFromRecordBeginningToCurrentHour(target[0]);
        expect(wrapper).toMatchSnapshot();
      });
    });
    describe("getHoursFromRecordBeginningToCurrentHour Helper", () => {
      it("With Correct params, should match snapshot", () => {
        const wrapper = helpers.getHoursFromRecordBeginningToCurrentHour();
        expect(wrapper).toMatchSnapshot();
      });
      it("After midnight, should match snapshot", () => {
        jest.resetModules();
        jest.doMock("moment", () => () => ({"hour": () => 5}));
        const {getHoursFromRecordBeginningToCurrentHour} = require("../helpers");
        const wrapper = getHoursFromRecordBeginningToCurrentHour();
        expect(wrapper).toMatchSnapshot();
      });
    });
    describe("getGravityClassName Helper", () => {
      it("With value under target, should return danger className", () => {
        const wrapper = helpers.getGravityClassName(actualTarget1[0], value);
        expect(wrapper).toMatchSnapshot();
      });
      it("With value superior to target, should return ok className", () => {
        const wrapper = helpers.getGravityClassName(actualTarget2[0], value);
        expect(wrapper).toMatchSnapshot();
      });
    });
    describe("getGapPercent Helper", () => {
      it("With Correct params, should match snapshot", () => {
        const wrapper = helpers.getGapPercent(target[0], value);
        expect(wrapper).toMatchSnapshot();
      });
    });
    describe("getGapBetweenActualAndTargetValue Helper", () => {
      it("With Correct params, should match snapshot", () => {
        const wrapper = helpers.getGapBetweenActualAndTargetValue(target[0], value);
        expect(wrapper).toMatchSnapshot();
      });
    });
    describe("getActualTarget Helper", () => {
      it("With Correct params, should match snapshot", () => {
        const wrapper = helpers.getActualTarget(target[0]);
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
  describe("CumulativeIndicatorChart Helpers", () => {
    beforeEach(() => {
      jest.resetModules();
      jest.dontMock("moment");
      helpers = require("../helpers");
      moment = require("moment");
    });
    describe("getYesterday Helper", () => {
      it("Should return yesterday date", () => {
        const expected = moment().
          subtract(1, "days").
          format("YYYYMMDDHH");
        const actual = helpers.
          getYesterday(target[0]).
          format("YYYYMMDDHH");
        expect(actual).toEqual(expected);
      });
    });
    describe("getFirstDayOfThisMonth Helper", () => {
      it("Should return the first day date of the given month", () => {
        const date = moment([
          2019,
          3,
          14
        ]);
        const expected = moment([
          2019,
          3,
          1
        ]).format("YYYYMMDD");
        const actual = helpers.getFirstDayOfThisMonth(date).format("YYYYMMDD");
        expect(actual).toEqual(expected);
      });
    });
    describe("getDaysOfMonth Helper", () => {
      it("With Correct params, should match snapshot", () => {
        const date = moment([
          2019,
          3,
          14
        ]);
        const actual = helpers.getDaysOfMonth(date);
        expect(actual).toMatchSnapshot();
      });
    });
    describe("getTargetsFromRecordBeginningToCurrentDay Helper", () => {
      it("With Correct params, should match snapshot", () => {
        const date = moment([
          2019,
          3,
          14
        ]);
        const actual = helpers.getTargetsFromRecordBeginningToCurrentDay(monthlyTarget[0], date);
        expect(actual).toMatchSnapshot();
      });
    });
    describe("CumulativeIndicatorChart Helpers", () => {
      beforeEach(() => {
        const date = moment([
          2019,
          2,
          19
        ]);
        jest.resetModules();
        jest.doMock("moment", () => () => (date.clone()));
        helpers = require("../helpers");
        moment = require("moment");
      });
      describe("mapMonthlyChartData Helper", () => {
        it("With Correct params, should match snapshot", () => {
          const wrapper = helpers.mapMonthlyChartData({
            "title": "title",
            "subtitle": "subtitle",
            "unit": "unit",
            "target": monthlyTarget,
            values
          });
          expect(wrapper).toMatchSnapshot();
        });
      });
      describe("mapChartData Helper", () => {
        it("With Daily timeline, should call mapDailyChartData", () => {
          const mapChartParams = {
            "title": "title",
            "subtitle": "subtitle",
            "unit": "unit",
            target,
            values,
            "timeLineType": TIMELINETYPE.DAILY
          };
          const wrapper = helpers.mapChartData(mapChartParams);
          expect(wrapper).toMatchSnapshot();
        });
        it("With Monthly timeline, should call mapMonthlyChartData", () => {
          const mapChartParams = {
            "title": "title",
            "subtitle": "subtitle",
            "unit": "unit",
            "target": monthlyTarget,
            values,
            "timeLineType": TIMELINETYPE.MONTHLY
          };
          const wrapper = helpers.mapChartData(mapChartParams);
          expect(wrapper).toMatchSnapshot();
        });
        it("With Yearly timeline, should call mapYearlyChartData", () => {
          jest.resetModules();
          jest.dontMock("moment");
          moment = require("moment");
          const date = moment([
            2019,
            10,
            8
          ]);
          jest.doMock("moment", () => () => (date));
          helpers = require("../helpers");
          moment = require("moment");
          const mapChartParams = {
            "title": "title",
            "subtitle": "subtitle",
            "unit": "unit",
            "target": yearlyTarget,
            values,
            "timeLineType": TIMELINETYPE.YEARLY
          };
          const wrapper = helpers.mapChartData(mapChartParams);
          expect(wrapper).toMatchSnapshot();
        });
      });
    });
    describe("getMonthsOfThisYear Helper", () => {
      it("With Correct params, should match snapshot", () => {
        const date = moment([
          2019,
          9,
          8
        ]);
        jest.resetModules();
        jest.doMock("moment", () => () => ({
          ...date,
          "month": (month) => ((month && date.clone().month(month)) || date.clone())
        }));
        helpers = require("../helpers");
        moment = require("moment");
        const actual = helpers.getMonthsOfThisYear(date);
        expect(actual).toMatchSnapshot();
      });
    });
    describe("getTargetsFromRecordBeginningToCurrentMonth Helper", () => {
      it("With Correct params, should match snapshot", () => {
        const actual = yearlyTarget;
        expect(actual).toMatchSnapshot();
      });
    });
    describe("mapValuesToCumulativeValues Helper", () => {
      it("With Correct params, should match snapshot", () => {
        const expected = [
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8
        ];
        const actual = helpers.mapValuesToCumulativeValues([
          1,
          1,
          1,
          1,
          1,
          1,
          1,
          1
        ]);
        expect(actual).toEqual(expected);
      });
    });
  });
});
