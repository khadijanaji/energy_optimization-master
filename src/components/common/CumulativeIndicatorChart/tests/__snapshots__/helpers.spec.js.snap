// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component CumulativeIndicatorChart Helpers CumulativeIndicatorChart Helpers mapChartData Helper With Daily timeline, should call mapDailyChartData 1`] = `
Object {
  "chart": Object {
    "pinchType": "xy",
    "type": "spline",
    "zoomType": "xy",
  },
  "plotOptions": Object {
    "line": Object {
      "dataLabels": Object {
        "enabled": true,
      },
      "enableMouseTracking": false,
    },
  },
  "series": Array [
    Object {
      "color": "#A8FBA8",
      "dashStyle": "Dot",
      "data": Array [
        0,
        6,
        13,
        19,
        25,
        31,
        38,
        44,
        50,
        56,
        63,
        69,
        75,
        81,
        88,
        94,
        100,
      ],
      "name": "Objectif",
    },
    Object {
      "data": Array [
        0,
        10,
        30,
        50,
        78.5,
        96.17,
        123.45,
      ],
      "name": "Valeurs actuelles",
    },
  ],
  "subtitle": Object {
    "text": "subtitle",
  },
  "title": Object {
    "text": "title",
  },
  "tooltip": Object {
    "formatter": [Function],
    "useHTML": true,
  },
  "xAxis": Object {
    "categories": Array [
      "7 h",
      "8 h",
      "9 h",
      "10 h",
      "11 h",
      "12 h",
      "13 h",
      "14 h",
      "15 h",
      "16 h",
      "17 h",
      "18 h",
      "19 h",
      "20 h",
      "21 h",
      "22 h",
      "23 h",
      "0 h",
    ],
  },
  "yAxis": Object {
    "title": Object {
      "text": "unit",
    },
  },
}
`;

exports[`Component CumulativeIndicatorChart Helpers CumulativeIndicatorChart Helpers mapChartData Helper With Monthly timeline, should call mapMonthlyChartData 1`] = `
Object {
  "chart": Object {
    "pinchType": "xy",
    "type": "spline",
    "zoomType": "xy",
  },
  "plotOptions": Object {
    "line": Object {
      "dataLabels": Object {
        "enabled": true,
      },
      "enableMouseTracking": false,
    },
  },
  "series": Array [
    Object {
      "color": "#A8FBA8",
      "dashStyle": "Dot",
      "data": Array [
        4161,
        8323,
        12484,
        16645,
        20806,
        24968,
        29129,
        33290,
        37452,
        41613,
        45774,
        49935,
        54097,
        58258,
        62419,
        66581,
        70742,
        74903,
      ],
      "name": "Objectif",
    },
    Object {
      "data": Array [
        10,
        40,
        90,
        168.5,
        264.67,
        388.12,
      ],
      "name": "Valeurs actuelles",
    },
  ],
  "subtitle": Object {
    "text": "subtitle",
  },
  "title": Object {
    "text": "title",
  },
  "tooltip": Object {
    "formatter": [Function],
    "useHTML": true,
  },
  "xAxis": Object {
    "categories": Array [
      "01",
      "02",
      "03",
      "04",
      "05",
      "06",
      "07",
      "08",
      "09",
      "10",
      "11",
      "12",
      "13",
      "14",
      "15",
      "16",
      "17",
      "18",
    ],
  },
  "yAxis": Object {
    "title": Object {
      "text": "unit",
    },
  },
}
`;

exports[`Component CumulativeIndicatorChart Helpers CumulativeIndicatorChart Helpers mapChartData Helper With Yearly timeline, should call mapYearlyChartData 1`] = `
Object {
  "chart": Object {
    "pinchType": "xy",
    "type": "column",
    "zoomType": "xy",
  },
  "plotOptions": Object {
    "line": Object {
      "dataLabels": Object {
        "enabled": true,
      },
      "enableMouseTracking": false,
    },
  },
  "series": Array [
    Object {
      "color": "#A8FBA8",
      "dashStyle": "Dot",
      "data": Array [
        1350,
        3000,
        5012,
        7850,
        9617,
        12345,
      ],
      "name": "Objectif",
    },
    Object {
      "data": Array [
        10,
        30,
        50,
        78.5,
        96.17,
        123.45,
      ],
      "name": "Valeurs actuelles",
    },
  ],
  "subtitle": Object {
    "text": "subtitle",
  },
  "title": Object {
    "text": "title",
  },
  "tooltip": Object {
    "formatter": [Function],
    "useHTML": true,
  },
  "xAxis": Object {
    "categories": Array [
      "Jan",
      "Feb",
      "Mar",
      "Apr",
      "May",
      "Jun",
      "Jul",
      "Aug",
      "Sep",
      "Oct",
      "Nov",
    ],
  },
  "yAxis": Object {
    "title": Object {
      "text": "unit",
    },
  },
}
`;

exports[`Component CumulativeIndicatorChart Helpers CumulativeIndicatorChart Helpers mapMonthlyChartData Helper With Correct params, should match snapshot 1`] = `
Object {
  "chart": Object {
    "pinchType": "xy",
    "type": "spline",
    "zoomType": "xy",
  },
  "plotOptions": Object {
    "line": Object {
      "dataLabels": Object {
        "enabled": true,
      },
      "enableMouseTracking": false,
    },
  },
  "series": Array [
    Object {
      "color": "#A8FBA8",
      "dashStyle": "Dot",
      "data": Array [
        4161,
        8323,
        12484,
        16645,
        20806,
        24968,
        29129,
        33290,
        37452,
        41613,
        45774,
        49935,
        54097,
        58258,
        62419,
        66581,
        70742,
        74903,
      ],
      "name": "Objectif",
    },
    Object {
      "data": Array [
        10,
        40,
        90,
        168.5,
        264.67,
        388.12,
      ],
      "name": "Valeurs actuelles",
    },
  ],
  "subtitle": Object {
    "text": "subtitle",
  },
  "title": Object {
    "text": "title",
  },
  "tooltip": Object {
    "formatter": [Function],
    "useHTML": true,
  },
  "xAxis": Object {
    "categories": Array [
      "01",
      "02",
      "03",
      "04",
      "05",
      "06",
      "07",
      "08",
      "09",
      "10",
      "11",
      "12",
      "13",
      "14",
      "15",
      "16",
      "17",
      "18",
    ],
  },
  "yAxis": Object {
    "title": Object {
      "text": "unit",
    },
  },
}
`;

exports[`Component CumulativeIndicatorChart Helpers getActualTarget Helper With Correct params, should match snapshot 1`] = `75.1`;

exports[`Component CumulativeIndicatorChart Helpers getDaysOfMonth Helper With Correct params, should match snapshot 1`] = `
Array [
  "01",
  "02",
  "03",
  "04",
  "05",
  "06",
  "07",
  "08",
  "09",
  "10",
  "11",
  "12",
  "13",
  "14",
]
`;

exports[`Component CumulativeIndicatorChart Helpers getGapBetweenActualAndTargetValue Helper With Correct params, should match snapshot 1`] = `-26.749999999999986`;

exports[`Component CumulativeIndicatorChart Helpers getGapPercent Helper With Correct params, should match snapshot 1`] = `-17.809587217043934`;

exports[`Component CumulativeIndicatorChart Helpers getGravityClassName Helper With value superior to target, should return ok className 1`] = `"cumulative-indicator-chart__container__value__OK"`;

exports[`Component CumulativeIndicatorChart Helpers getGravityClassName Helper With value under target, should return danger className 1`] = `"cumulative-indicator-chart__container__value__DANGER"`;

exports[`Component CumulativeIndicatorChart Helpers getHoursFromRecordBeginningToCurrentHour Helper After midnight, should match snapshot 1`] = `
Array [
  "7 h",
  "8 h",
  "9 h",
  "10 h",
  "11 h",
  "12 h",
  "13 h",
  "14 h",
  "15 h",
  "16 h",
  "17 h",
  "18 h",
  "19 h",
  "20 h",
  "21 h",
  "22 h",
  "23 h",
  "0 h",
  "1 h",
  "2 h",
  "3 h",
  "4 h",
  "5 h",
]
`;

exports[`Component CumulativeIndicatorChart Helpers getHoursFromRecordBeginningToCurrentHour Helper With Correct params, should match snapshot 1`] = `
Array [
  "7 h",
  "8 h",
  "9 h",
  "10 h",
  "11 h",
  "12 h",
  "13 h",
  "14 h",
  "15 h",
  "16 h",
  "17 h",
  "18 h",
  "19 h",
]
`;

exports[`Component CumulativeIndicatorChart Helpers getMonthsOfThisYear Helper With Correct params, should match snapshot 1`] = `
Array [
  "Jan",
  "Feb",
  "Mar",
  "Apr",
  "May",
  "Jun",
  "Jul",
  "Aug",
  "Sep",
  "Oct",
]
`;

exports[`Component CumulativeIndicatorChart Helpers getTargetsFromRecordBeginningToCurrentDay Helper With Correct params, should match snapshot 1`] = `
Array [
  4300,
  8600,
  12900,
  17200,
  21500,
  25800,
  30100,
  34400,
  38700,
  43000,
  47300,
  51600,
  55900,
  60200,
]
`;

exports[`Component CumulativeIndicatorChart Helpers getTargetsFromRecordBeginningToCurrentHour Helper With Correct params, should match snapshot 1`] = `
Array [
  0,
  6,
  13,
  19,
  25,
  31,
  38,
  44,
  50,
  56,
  63,
  69,
  75,
]
`;

exports[`Component CumulativeIndicatorChart Helpers getTargetsFromRecordBeginningToCurrentMonth Helper With Correct params, should match snapshot 1`] = `
Array [
  1350,
  3000,
  5012,
  7850,
  9617,
  12345,
]
`;

exports[`Component CumulativeIndicatorChart Helpers getTodayProductionHours Helper With Correct params, should match snapshot 1`] = `13`;

exports[`Component CumulativeIndicatorChart Helpers mapDailyChartData Mapper With Correct params, should match snapshot 1`] = `
Object {
  "chart": Object {
    "pinchType": "xy",
    "type": "spline",
    "zoomType": "xy",
  },
  "plotOptions": Object {
    "line": Object {
      "dataLabels": Object {
        "enabled": true,
      },
      "enableMouseTracking": false,
    },
  },
  "series": Array [
    Object {
      "color": "#A8FBA8",
      "dashStyle": "Dot",
      "data": Array [
        0,
        6,
        13,
        19,
        25,
        31,
        38,
        44,
        50,
        56,
        63,
        69,
        75,
      ],
      "name": "Objectif",
    },
    Object {
      "data": Array [
        0,
        10,
        30,
        50,
        78.5,
        96.17,
        123.45,
      ],
      "name": "Valeurs actuelles",
    },
  ],
  "subtitle": Object {
    "text": "subtitle",
  },
  "title": Object {
    "text": "title",
  },
  "tooltip": Object {
    "formatter": [Function],
    "useHTML": true,
  },
  "xAxis": Object {
    "categories": Array [
      "7 h",
      "8 h",
      "9 h",
      "10 h",
      "11 h",
      "12 h",
      "13 h",
      "14 h",
      "15 h",
      "16 h",
      "17 h",
      "18 h",
      "19 h",
    ],
  },
  "yAxis": Object {
    "title": Object {
      "text": "unit",
    },
  },
}
`;

exports[`Component CumulativeIndicatorChart Helpers renderValue Render With correct data, should render value 1`] = `"123.450"`;

exports[`Component CumulativeIndicatorChart Helpers renderValue Render With incorrect value, should render N/A 1`] = `
<span
  className="cumulative-indicator-chart__container__value__UNAVAILABLE"
>
  Donnée non disponible
</span>
`;

exports[`Component CumulativeIndicatorChart Helpers renderValue Render With loading status, should render loader 1`] = `
<img
  alt="chargement"
  className="category-indicator-item__loading"
  src=""
/>
`;
